### PR TITLE
ENH: Allow early stopping through scikit-learn API

### PR DESCRIPTION
**Pull request by jseabold**
Don't merge yet but feel free to review. I want to get a little more experience using it. 

Things I'm not sure about. 

Should `sample_weights` also be passed through to the `evals` DMatrices? I also didn't change either of the `predict` functions to use the `best_iteration_` number from early-stopping mainly because this is not usually want I'd want to do, but I'm open to being convinced that it would be helpful/desirable.

### CODE REVIEW

**Comment by jseabold** [comment]
TODO: it would be nice to be have the ability to train on both training and evaluation after early stopping.

**Comment by tqchen** [comment]
Let us keep predict as it is instead of using best_iteration.

 I agree that it is interesting to make eval_metric and feval as one parameter, similarly for fobj and objective

**Comment by tqchen** [comment]
This seems to be a complete PR, please let me know if you want to merge it in

**Comment by jseabold** [comment]
I want to take one more pass at fixing the API if you don't mind. Will get
to it in the next few days.

**Comment by tqchen** [comment]
Good, thanks

**Comment by tqchen** [comment]
@jseabold  do you have any updates on this

**Comment by jseabold** [comment]
Am out of the country at the moment, but haven't forgotten about it and
will revisit soon. Would really like cleaner handling of the custom
functions before we're tied to this as the API.

**Comment by tqchen** [comment]
Great, looking forward toit

**Comment by jseabold** [comment]
Updated.

**Comment by tqchen** [comment]
- I see your changes in the customized evaluation function. Are you going to change objective as well?
- It would be great to move these changes to the xgb.train function, so both sklearn interface and xgboost  function can benefit from it

**Comment by jseabold** [comment]
I was worried about backwards compatibility in the `train` function. What's your deprecation policy?

**Comment by jseabold** [comment]
What do you think about eventually dropping (slowly and politely) the `xgboost` interface for the `sklearn` interface? (Not in this PR, but so things don't have to be maintained twice.)

**Comment by tqchen** [comment]
- For backward compatibility let us keep the fobj and feval, but also detect the callale in objective and eval_metric parameters.
   - Update the examples according to preferrable directly use the objective and eval metric
- I think we are OK to break backward compatibility to improve the advanced features.

- I think we want to keep the ```train``` interface, for a few reasons
    - Consistency across all languages.
    - Since most of calls of sklearn calls redirects into ```train``` and there won't be too much cost in maintaining the ```train``` (otherwise code goes into sklearn interface)
    - The ```xgboost``` interface makes it easier to experiment with new parameters, etc, with param as arbitrary dict
    - The  possible support for visualization module that demand more features beyond sklearn

This being said, I think we have learnt something in the past and we are happy to discuss what is a better interface we want to cover flexibility and usibility. We can open an issue to track and discuss this.

Thanks!

**Comment by tqchen** [comment]
@jseabold I just pushed in some change of styles that may globally affect the files. The reason was mainly mentioned in #372 for better future checks. 
It seems make your current PR not auto-mergable, can you pull from the latest version before we final merge it in? Thanks!

I am hoping the new CI can have more checks in the tool and we will be safer making changes. If you have any suggestion (e.g. how we should tweak the linter to disable some warnings, please let me know)

**Comment by jseabold** [comment]
Rebased.

**Comment by tqchen** [comment]
seems there are some checks failures, can you take  a look? https://travis-ci.org/dmlc/xgboost/builds/69555960

**Comment by jseabold** [comment]
Sorry rebase victim. Let's see how this looks.

**Comment by jseabold** [comment]
Travis failure looks unrelated to this PR.

**Comment by tqchen** [comment]
Nop, seems that was an R problem, which was a bit strange. I can go head and merge it in. But it would be great if you can move the check of callable the feval and fobjective into xgb.train so all the interface can benefit from the change.

Thanks for the hard work!

**Comment by tqchen** [comment]
@jseabold  Do you have any comments on this. Let me know if you like to move the evalset logic to the train part and reuse that or we could make that as separate PR

Thanks!

**Comment by tqchen** [comment]
OK, I am merge this into the master for now, since this one is already self-contained. Thanks for great effort!
